cpp/include/raft/matrix/detail/init.cuh

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace raft::matrix::detail {
+
+/**
+ * @brief Create a diagonal identity matrix
+ * @param matrix: matrix of size total_size
+ * @param lda: Leading dimension
+ * @param total_size: number of elements of the matrix
+ */
+template <typename m_t, typename idx_t = int>
```

Comment:
What do you think about moving this to `matrix/detail/diagonal.cuh` and using `set_diagonal()`, along w/ `cudaMemsetAsync` (or `matrix::fill` to fill the matrix w/ 0s)? It would consolidate this so we don't have to introduce new kernels.